File: leetcode/java/0112_path_sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int currSum = 0;
    
    public boolean hasPathSum(TreeNode root, int targetSum) {
        // helper(node, targetSum, currSum) -> returns true if node is a leaf node and targetSum == currSum
    
        if (root != null) {
            currSum += root.val;
            if (hasPathSum(root.left, targetSum) == true) return true;

            if (currSum == targetSum && (root.left == null && root.right == null)) return true;

            if (hasPathSum(root.right, targetSum) == true) return true;
            currSum -= root.val;
        }

        return false;
    }
}

File: leetcode/cpp/0112_path_sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool sumHelper(TreeNode* root, int targetSum, int& currSum) {
        if (root != nullptr) {
            currSum += root->val;
            if (sumHelper(root->left, targetSum, currSum) == true) {
                return true;
            }

            if (root->left == nullptr && root->right == nullptr && currSum == targetSum) {
                return true;
            }

            if (sumHelper(root->right, targetSum, currSum) == true) {
                return true;
            }
            currSum -= root->val;
        }

        return false;
    }
    
    bool hasPathSum(TreeNode* root, int targetSum) {
        int currSum = 0;

        return sumHelper(root, targetSum, currSum);
    }
};